File: Pythoncode/39.组合总和.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        res = []
        self.helper(candidates, target, 0, [], res)
        return res
    def helper(self, candidates, target, sum_so_far, candidates_so_far, res):
        if sum_so_far == target:
            res.append(candidates_so_far)
        for i in range(len(candidates)):
            n = candidates[i]
            if sum_so_far + n > target:
                break
            else:
                new_candidates_so_far = candidates_so_far[:]
                new_candidates_so_far.append(n)
                self.helper(candidates[i:], target, sum_so_far+n, new_candidates_so_far, res)

File: Cppcode/39.组合总和.cpp
#include <vector>
#include <numeric>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        reverse(candidates.begin(), candidates.end());
        vector<vector<int>> res;
        vector<int> resEntry;
        helper(res, resEntry, candidates, target);
        return res;
    }

    void helper(vector<vector<int>>& res, vector<int> resEntry, vector<int> candidates, int target)
    {
        if (candidates.empty()) return;
        while (true)
        {
            auto s = accumulate(resEntry.begin(), resEntry.end(), 0);
            if (s < target)
            {
                helper(res, resEntry, vector<int>(candidates.begin()+1, candidates.end()), target);
            }
            else if (s == target)
            {
                res.push_back(resEntry);
            }
            else
            {
                break;
            }
            resEntry.push_back(candidates[0]);
        }
        
    }
};